File: fibonacci-dotnet/src/Fibonacci.Api/Algorithms/IterativeAlgorithm.cs
using System.Numerics;
using Fibonacci.Api.Constants;

namespace Fibonacci.Api.Algorithms;

/// <summary>
/// Iterative Fibonacci — O(n) time, O(1) space.
/// Optimal for moderate n: no stack overhead, no cache, minimal memory footprint.
/// </summary>
public sealed class IterativeAlgorithm : IFibonacciAlgorithm
{
    public string Name => "iterative";
    public string Label => "Iterative";
    public string TimeComplexity => "O(n)";
    public string SpaceComplexity => "O(1)";
    public int MaxN => FibonacciConstants.MaxN;
    public bool RequiresExplicitInclusion => false;

    public BigInteger Compute(int n)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(n);

        BigInteger a = BigInteger.Zero, b = BigInteger.One;
        for (var i = 0; i < n; i++)
            (a, b) = (b, a + b);

        return a;
    }
}

File: fibonacci-java/src/main/java/com/bretzelcoder/fibonacci/algorithm/IterativeAlgorithm.java
package com.bretzelcoder.fibonacci.algorithm;

import com.bretzelcoder.fibonacci.config.FibonacciConstants;
import java.math.BigInteger;
import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Component;

/**
 * Bottom-up iteration — O(n) time, O(1) space.
 * Holds only the last two values: no recursion stack, no cache, no allocation beyond
 * the BigInteger arithmetic itself. Usually the fastest of the four at moderate n.
 */
@Component
@Order(3)
public class IterativeAlgorithm implements FibonacciAlgorithm {

    @Override
    public String name() {
        return "iterative";
    }

    @Override
    public String label() {
        return "Iterative";
    }

    @Override
    public String timeComplexity() {
        return "O(n)";
    }

    @Override
    public String spaceComplexity() {
        return "O(1)";
    }

    @Override
    public int maxN() {
        return FibonacciConstants.MAX_N;
    }

    @Override
    public boolean requiresExplicitInclusion() {
        return false;
    }

    @Override
    public BigInteger compute(int n) {
        requireNonNegative(n);

        BigInteger a = BigInteger.ZERO;
        BigInteger b = BigInteger.ONE;
        for (int i = 0; i < n; i++) {
            BigInteger next = a.add(b);
            a = b;
            b = next;
        }
        return a;
    }
}
